forge-ai/src/main/java/forge/ai/PlayerControllerAi.java

```diff
@@ -105,8 +102,98 @@ public boolean isAI() {
 
     @Override
     public List<PaperCard> sideboard(Deck deck, GameType gameType, String message) {
-        // AI does not know how to sideboard
-        return null;
+        if (!getAi().getBooleanProperty(AiProps.SIDEBOARDING_ENABLE)) {
+            return null;
+        }
+
+        boolean sbLimitedFormats = getAi().getBooleanProperty(AiProps.SIDEBOARDING_IN_LIMITED_FORMATS);
+        boolean sbSharedTypesOnly = getAi().getBooleanProperty(AiProps.SIDEBOARDING_SHARED_TYPE_ONLY);
+        boolean sbPlaneswalkerException = getAi().getBooleanProperty(AiProps.SIDEBOARDING_PLANESWALKER_EQ_CREATURE);
+        int sbChanceOnWin = getAi().getIntProperty(AiProps.SIDEBOARDING_CHANCE_ON_WIN);
+        int sbChancePerCard = getAi().getIntProperty(AiProps.SIDEBOARDING_CHANCE_PER_CARD);
+
+        if (!sbLimitedFormats && gameType.isCardPoolLimited()) {
+            return null;
+        }
+
+        if (MyRandom.getRandom().nextInt(100) < sbChanceOnWin) {
+            GameOutcome lastOutcome = brains.getGame().getMatch().getOutcomeById(brains.getGame().getId() - 1);
+            if (lastOutcome.getWinningPlayer().getPlayer().equals(player.getLobbyPlayer())) {
+                return null;
+            }
+        }
+
+        List<PaperCard> main = deck.get(DeckSection.Main).toFlatList();
+        List<PaperCard> sideboard = deck.get(DeckSection.Sideboard).toFlatList();
+
+        // Devise a sideboarding plan
+        // TODO: Allow a pre-planned (e.g. fully transformative) sideboard via deck metadata of some kind?
+        List<PaperCard> processed = Lists.newArrayList();
+        Map<PaperCard, PaperCard> sideboardPlan = Maps.newHashMap();
+        for (PaperCard cSide : sideboard) {
+            if (processed.contains(cSide)) {
+                continue;
+            } else if (cSide.getRules().getAiHints().getRemAIDecks()) {
+                continue; // don't sideboard in anything that we don't know how to play
+            } else if (cSide.getRules().getType().isLand()) {
+                continue; // don't know how to sideboard lands efficiently yet
+            }
+
+            for (PaperCard cMain : main) {
+                if (processed.contains(cMain)) {
+                    continue;
+                } else if (cMain.equals(cSide)) {
```

Comment:
is this check so it skips identical cards?
then it probably needs to compare by name instead?